### PR TITLE
Use `clang-19`

### PR DESCRIPTION
**Pull request by thevar1able**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 5eb5d420d6f6cfdbda69992ca36c280686821067 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/70414/5eb5d420d6f6cfdbda69992ca36c280686821067/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/70320/7e7a92cef6bdd48dca06cf221c91a813dcb757e8/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: cmake/clang_tidy.cmake [hunk 1]
```diff
@@ -5,14 +5,14 @@ if (ENABLE_CLANG_TIDY)
 
     find_program (CLANG_TIDY_CACHE_PATH NAMES "clang-tidy-cache")
     if (CLANG_TIDY_CACHE_PATH)
-        find_program (_CLANG_TIDY_PATH NAMES "clang-tidy-18" "clang-tidy-17" "clang-tidy-16" "clang-tidy")
+        find_program (_CLANG_TIDY_PATH NAMES "clang-tidy-19" "clang-tidy-18" "clang-tidy-17" "clang-tidy-16" "clang-tidy")
```
Comment (rschu1ze) [inline cmake/clang_tidy.cmake:8]: We usually remain compatible with the past three Clang (-tidy) versions. Therefore, remove clang-tidy 17?

**Review by rschu1ze — COMMENTED** [review]
Could you please also check https://github.com/ClickHouse/ClickHouse/pull/60469 about other files that need to be changed?

**Comment by rschu1ze** [comment]
Include https://github.com/ClickHouse/grpc/pull/42 as well?

**Comment by thevar1able** [comment]
>Include https://github.com/ClickHouse/grpc/pull/42 in this PR as well?

See https://github.com/ClickHouse/ClickHouse/pull/70412

File: cmake/clang_tidy.cmake (hunk 1 above)
Comment (thevar1able) [reply cmake/clang_tidy.cmake:8]: 16*, :ok_hand:

**Comment by thevar1able** [comment]
TODO
```
/home/thevar1able/testmount/ClickHouse/src/Common/HashTable/HashTable.h:174:66: error: returning a constant reference parameter may cause use-after-free when the parameter is constructed from a temporary [bugprone-return-const-ref-from-parameter,-warnings-as-errors]
  174 |     static const Key & getKey(const value_type & value) { return value; }
      |                                                                  ^~~~~
/home/thevar1able/testmount/ClickHouse/src/Common/HashTable/HashTableKeyHolder.h:53:65: error: returning a constant reference parameter may cause use-after-free when the parameter is constructed from a temporary [bugprone-return-const-ref-from-parameter,-warnings-as-errors]
   53 | inline Key & ALWAYS_INLINE keyHolderGetKey(Key && key) { return key; }
      |                                                                 ^~~
```

**Comment by thevar1able** [comment]
@rschu1ze `clang-tidy-19` now complains about not modernizing ranges; should we apply its suggestions, or add some ignores to conifg?

```
-boost-use-ranges
-modernize-use-ranges
```

**Comment by rschu1ze** [comment]
> @rschu1ze `clang-tidy-19` now complains about not modernizing ranges; should we apply its suggestions, or add some ignores to conifg?
> 
> ```
> -boost-use-ranges
> -modernize-use-ranges
> ```

@thevar1able Most colleagues are still unfamiliar with C++20 ranges, also people on the internet have various complains about ranges (funny syntax, slow, etc.). I'd prefer to not force ranges on everyone and add ignores instead.

File: .clang-tidy [hunk 2]
```diff
@@ -90,6 +93,7 @@ Checks: [
   '-misc-non-private-member-variables-in-classes',
   '-misc-confusable-identifiers', # useful but slooo
   '-misc-use-anonymous-namespace',
+  '-misc-use-internal-linkage',
```
Comment (rschu1ze) [inline .clang-tidy:96]: Based on the [docs](https://clang.llvm.org/extra/clang-tidy/checks/misc/use-internal-linkage.html), this warning seems useful. If there is a reasonable number of occurrences (e.g. < 20), it IMHO makes sense to have the warning enabled.

File: .clang-tidy [hunk 3]
```diff
@@ -137,6 +141,7 @@ Checks: [
   '-readability-suspicious-call-argument',
   '-readability-uppercase-literal-suffix',
   '-readability-use-anyofallof',
+  '-readability-math-missing-parentheses',
```
Comment (rschu1ze) [inline .clang-tidy:144]: Althought this warning seems a bit controversial, I still like it. If there are not tons of occurrences, consider enabling it (up to you).

Comment (rschu1ze) [inline .clang-tidy:144]: Althought this warning seems a bit controversial, I still like it. If there are not tons of occurrences, consider enabling it (up to you).

**Comment by rschu1ze** [comment]
@thevar1able Perhaps you also like to have a look at #66625. This was an earlier attempt to make clang-tidy 19 work. I think that both `bugprone-*` warnings are potentially useful but, again, only if there are not too many occurrences.

File: .clang-tidy (hunk 2 above)
Comment (thevar1able) [reply .clang-tidy:96]: At the moment it looks like a ton of errors for all the rules I've disabled; I'll recheck, I want to disable as little `tidy` rules as possible, but at the same time I want to keep this PR small and readable. I may do another `clang-tidy -fix` pass, but it already broke our code (in that huge PR) to the point of subtle errors caught only by ASAN.

**Comment by Felixoid** [comment]
@thevar1able, there's an issue related to the sccache failure https://github.com/ClickHouse/ClickHouse/issues/68266 with a probably related upstream issue https://github.com/mozilla/sccache/issues/1153

Let us invest some time into fixing the sccache by adding either retries or ignores for errors.

From the retries perspective, well, we have a hacky way to restart any job [with fire](https://github.com/ClickHouse/ClickHouse/blob/master/tests/ci/ci_utils.py#L44)

**Comment by thevar1able** [comment]
Thank you, @Felixoid. Let's continue the discussion [here](https://github.com/ClickHouse/ClickHouse/pull/70522#discussion_r1795091688).

**Review by rschu1ze — APPROVED** [review]

File: docker/test/fasttest/requirements.txt [hunk 4]
```diff
@@ -27,7 +27,7 @@ pandas==1.5.3
 pip==24.1.1
 pipdeptree==2.23.0
 pyparsing==2.4.7
-python-apt==2.4.0+ubuntu3
+# python-apt==2.4.0+ubuntu3
```
Comment (thevar1able) [inline docker/test/fasttest/requirements.txt:30]: Probably useless, builds are passing without it.

Comment (thevar1able) [reply docker/test/fasttest/requirements.txt:30]: @Felixoid @alesapin maybe you know more?

Comment (Felixoid) [reply docker/test/fasttest/requirements.txt]: It comes from the older script version https://github.com/ClickHouse/ClickHouse/commit/ce4ddf73364d5e6877917f26a93890195bbe89e0

Yes, it's a system lib version, and can be ignored

File: src/Common/ZooKeeper/examples/zkutil_test_async.cpp [hunk 5]
```diff
@@ -17,6 +17,8 @@ try
 
     size_t num_threads = DB::parse<size_t>(argv[1]);
     std::vector<std::thread> threads;
+    threads.reserve(num_threads);
```
Comment (rschu1ze) [inline src/Common/ZooKeeper/examples/zkutil_test_async.cpp:20]: Does `.reserve` fix a clang-tidy warning? The thing is that the internal reserve (memory allocation) happens when the vector is created (l. 19) and at the latest in the first insert (l. 24). In all reasonable standard libraries, more space than the minimally needed space is reserved, so future inserts don't allocate again. Assuming that `num_threads` is "small", we will certainly allocate only once. So an explicit `.reserve` seems not really necessary.

File: .clang-tidy [hunk 6]
```diff
@@ -16,6 +16,9 @@ Checks: [
 
   '-android-*',
 
+  '-boost-use-ranges',
```
Comment (rschu1ze) [inline .clang-tidy:19]: Can't wait to get these two warnings disabled, they are currently spamming my local clang build like hell.

File: src/Common/ZooKeeper/examples/zkutil_test_async.cpp (hunk 5 above)
Comment (thevar1able) [reply src/Common/ZooKeeper/examples/zkutil_test_async.cpp:20]: >Does `.reserve` fix a clang-tidy warning?

Yes, it does.

File: src/Common/SignalHandlers.cpp [hunk 7]
```diff
@@ -634,7 +634,7 @@ void HandledSignals::setupCommonDeadlySignalHandlers()
 {
     /// SIGTSTP is added for debugging purposes. To output a stack trace of any running thread at anytime.
     /// NOTE: that it is also used by clickhouse-test wrapper
-    addSignalHandler({SIGABRT, SIGSEGV, SIGILL, SIGBUS, SIGSYS, SIGFPE, SIGPIPE, SIGTSTP, SIGTRAP}, signalHandler, true);
+    addSignalHandler({SIGABRT, SIGSEGV, SIGILL, SIGBUS, SIGSYS, SIGFPE, SIGTSTP, SIGTRAP}, signalHandler, true);
```
Comment (thevar1able) [inline src/Common/SignalHandlers.cpp:637]: https://github.com/llvm/llvm-project/pull/98200

Also, why do we block it [here](https://github.com/ClickHouse/ClickHouse/blob/3272f1143c42a93058dad167adbbcf2f12058654/src/Daemon/BaseDaemon.cpp#L436), but then add a handler?

**Comment by thevar1able** [comment]
Depends on https://github.com/llvm/llvm-project/issues/113385. The fix was merged and reverted.

**Comment by thevar1able** [comment]
The patch was merged, LLVM 19.1.4 is expected on Nov 19th https://discourse.llvm.org/t/llvm-19-1-4-release-update/83088

**Comment by thevar1able** [comment]
LLVM 19.1.4 was released https://github.com/llvm/llvm-project/releases/tag/llvmorg-19.1.4

**Pull request merged**